src/imars3dv2/filters/intensity_fluctuation_correction.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+import numpy as np
+import tomopy
+import tomopy.util.mproc as mproc
+from skimage import feature
+import concurrent.futures as cf
+
+
+def intensity_fluctuation_correction(
+    arrays: np.ndarray,
+    air_pixels: int = 5,
+    sigma: int = 3,
+    ncore: int = -1,
+):
+    """
+    Correct for intensity fluctuation in the radiograph.
+
+    Parameters
+    ----------
+    @param arrays:
+        The image/radiograph stack to correct for beam intensity fluctuation.
+    @param air_pixels:
+        Number of pixels at each boundary to calculate the scaling factor. When a negative number
+        is given, the auto air region detection will be used instead of tomopy.
+    @param sigma:
+        The standard deviation of the Gaussian filter, only valid when using the auto air region
+        detection via canny edge detection from skimage.
+    @param ncore:
+        The number of cores to use for parallel processing, default is -1, which means using all available cores.
+
+    Returns
+    -------
+    @return:
+        The corrected image/radiograph stack.
+    """
+    # validation
+    if arrays.ndim not in (2, 3):
+        raise ValueError("The image/radiograph stack must be 2D or 3D.")
+    # process
+    if air_pixels < 0:
+        # auto air region detection
+        ncore = mproc.mp.cpu_count() if ncore == -1 else int(ncore)
+        with cf.ProcessPoolExecutor(ncore) as e:
+            jobs = [
+                e.submit(
+                    intensity_fluctuation_correction_skimage,
+                    image,
+                    sigma,
+                )
+                for image in arrays
+            ]
+        return np.array([job.result() for job in jobs])
+    else:
+        # use tomopy process
+        ncore = None if ncore == -1 else int(ncore)
+        return tomopy.normalize_bg(arrays, air=air_pixels, ncore=ncore)
+
+
+def intensity_fluctuation_correction_skimage(
+    image: np.ndarray,
+    sigma: int = 3,
+):
+    """
+    Correct for intensity fluctuation in the radiograph using skimage to auto
+    detect the air region (adapted from iMars3dv1.filter.ifc).
+
+    Parameters
+    ----------
+    @param image:
+        The image/radiograph (2D) to correct for beam intensity fluctuation.
+    @param sigma:
+        The standard deviation of the Gaussian filter for the canny edge detection.
+
+    Returns
+    -------
+    @return:
+        The corrected image/radiograph (2D).
+    """
+    # get boundary
+    edge = feature.canny(image, sigma=sigma)
+    start_row = None
```

Comment:
I have mixed feelings on my advice here. Generally you don't want to change types to mark a special value. Would setting the special value to -1 be acceptable?
```suggestion
    start_row = -1  # indicates that the value has not been determined
```